### ISSUE TITLE
.dev.js files included production builds

### ISSUE THREAD

**Issue by jokeyrhyme** [redux-bootstrap/redux-bootstrap#43]
## Expected Behavior

Ideally, for `NODE_ENV=production` builds, we should produce builds that are as small and optimised as feasible.
## Current Behavior
- Code from redux-devtools can be found in build output
- As-is, a minimal react + redux-bootstrap build is ~400KB minified
- Manually dropping .dev.js files in redux-bootstrap results in build that is ~280KB
- Admittedly this only ends up with an extra ~120KB
- For a minimal app this is quite a waste, but this will likely be a much smaller fraction of a fully-featured complex app
## Possible Solution

Should we decide this is something that should be addressed, I'm not 100% sure how to achieve this. 
- If end up with no `require('....dev')` lines prior to build tools crawling the module dependency graph, then they won't be included at all.
- Or perhaps the build tools themselves can be configured to stub these out when `NODE_ENV=production`, in which case we could supply instructions on how to configure this.
## Steps to Reproduce (for bugs)
1. Create a fresh React project using https://github.com/facebookincubator/create-react-app
2. add redux-bootstrap as we instruct
3. `npm run build`
4. observe file sizes in the build directory
5. change node_modules/redux-bootstrap/lib/containers/root.js to avoid `require('./root.dev')`
6. change node_modules/redux-bootstrap/lib/store/configure_store.js to avoid `require('./configure_store.dev')`
7. `npm run build`
8. observe file sizes in the build directory
## Your Environment
- Version used: 1.0.0-rc.1
- Environment name and version (e.g. Chrome 39, node.js 5.4): node 6.3.1
- Operating System and version (desktop or mobile): ArchLinux
- Link to your project: https://github.com/jokeyrhyme/react-redux-bootstrap-example

**Comment by jokeyrhyme** [comment]
By the way, redux-bootstrap works terrifically with create-react-app. Very simple to combine.

I'll try to put a blog post up about it.

**Comment by remojansen** [comment]
Do you know what is used internally by `create-react-app` to create the build output? is it webpack or browserify?

**Comment by jokeyrhyme** [comment]
@remojansen looks like its webpack: https://github.com/facebookincubator/create-react-app/blob/master/package.json#L58

**Comment by jokeyrhyme** [comment]
Here's the example project: https://github.com/jokeyrhyme/react-redux-bootstrap-example

**Comment by remojansen** [comment]
Do you have access to the webpack config file? If so you should be able to configure the environment:

```js
plugins: [
  new webpack.DefinePlugin({
    'process.env': {
      'NODE_ENV': JSON.stringify('production')
    }
  })
],
```

Then the non-production modules should not be included in the build.

**Comment by jokeyrhyme** [comment]
@remojansen they already set that:

- https://github.com/facebookincubator/create-react-app/blob/v0.1.0/scripts/build.js#L10

- https://github.com/facebookincubator/create-react-app/blob/v0.1.0/config/webpack.config.prod.js#L113

I think the problem is that `NODE_ENV === 'production'` is a run-time check, where as `require('...')` has static implications during build. Build tools don't parse that run-time check, but they do honour the static imports.

**Comment by remojansen** [comment]
I don't really know how to fix it them :( I know in webpack you are allowed to do:
```
if(NODE_ENV === 'production') {
  import "A";
}  else {
  import "B";
}
```

But in typescript all imports must be in the top of the file (this is based on the ECMAScript standard):

```
 import * as A from "A";
 import * as B from "B";

if(NODE_ENV === 'production') {
  export { A };
}  else {
  export { B };
}
```

**Comment by jokeyrhyme** [comment]
Well, for webpack, maybe we can use this:

- https://webpack.github.io/docs/configuration.html#resolve-alias

We could alias the .dev.js files to the .prod.js files? This is still something the developer would have to do, but we provide instructions perhaps?

**Comment by jokeyrhyme** [comment]
It would be nicer if there was a more automatic or built-in way of doing this though. As the Facebook React CLI doesn't really let you mess with configuration without completely "ejecting" from all the nice automatic stuff it does for you.

**Comment by jokeyrhyme** [comment]
@remojansen I took a look at starter kit: https://github.com/davezuko/react-redux-starter-kit/blob/3fabc5f4/src/store/createStore.js#L17

It looks like they don't directly `require()` the dev tools module. They just hook into it if it is already present and globally-exposed. I'll keep looking for more examples of how to handle this.

**Comment by remojansen** [comment]
Hi the code sample that you have provided requires the users to have the redux-devtools chrome extension installed. Maybe we can ask for help at https://github.com/facebookincubator/create-react-app/issues/102 ?

**Comment by remojansen** [comment]
Hi @jokeyrhyme did you solve this problem? I think the solution is something like https://github.com/blacksonic/typescript-webpack-tree-shaking

**Comment by jokeyrhyme** [comment]
Tree shaking would be the answer, but it would require particularly thorough tree shaking. All conditions would need to be evaluated at build time, which seems more than what most people consider to be tree shaking.

My understanding is that tree shaking performs build-time static analysis of all `import` and `export` statements, determines which variables within each module are at all referenced by anything that leads to an `export`, and ditches all code that is unrelated.

I've personally just ditched the redux dock / panel and switched over completely to the devtools extension:
-   https://github.com/zalmoxisus/redux-devtools-extension#usage
-   https://github.com/blinkmobile/ketchum-all-ui/blob/master/src/redux/store.js

The extensions is a much better experience than the panel ever was, and it's actively maintained (unlike the panel).

**Comment by remojansen** [comment]
Thanks for that. I will try to migrate to the extension then 😄

**Comment by remojansen** [comment]
This should be fixed on 3.1.0 you need to use https://github.com/zalmoxisus/redux-devtools-extension to access the devtools.
